### PR TITLE
Fix completion for private items

### LINKED ISSUE #644: Don't suggest autocomplete for private functions
In the following code:

``` rust
mod my {
    // Items in modules default to private visibility.
    fn private_function() {
        println!("called `my::private_function()`");
    }
}

fn main() {
    my::
}
```

`intellij-rust` suggests `private_function` for autocomplete, but if a user were to select this, they get a compile error:

`error: function `private_function` is private`

I would think it would be best to omit private functions from the autocomplete.

There is also probably a corollary to `pub` vs. private `struct`s and `struct` members.

### PR DESCRIPTION
**Pull request by ortem**
Fixes #644

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/psi/ext/RsVisibility.kt [hunk 1]
```diff
@@ -22,3 +22,22 @@ fun RsVisibilityOwner.iconWithVisibility(flags: Int, icon: Icon): Icon =
         icon
     else
         icon.addVisibilityIcon(isPublic)
+
+fun RsVisible.isVisibleFrom(mod: RsMod): Boolean {
+    if (isPublic) return true
+
+    val elementMod = (if (this is RsMod) this.`super` else this.contextStrict()) ?: return true
+    // We have access to any item in any super module of `oMod`
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/psi/ext/RsVisibility.kt]: `oMod` is renamed to `mod` here. Please, fix comments

File: src/main/kotlin/org/rust/lang/core/resolve/ref/RsPathReferenceImpl.kt [hunk 2]
```diff
@@ -30,7 +30,13 @@ class RsPathReferenceImpl(
     override val RsPath.referenceAnchor: PsiElement get() = referenceNameElement
 
     override fun getVariants(): Array<out Any> =
-        collectCompletionVariants { processPathResolveVariants(ImplLookup.relativeTo(element), element, true, it) }
+        collectCompletionVariants {
+            processPathResolveVariants(ImplLookup.relativeTo(element),
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/resolve/ref/RsPathReferenceImpl.kt]: Move `ImplLookup.relativeTo(element)` to a next line

File: src/main/kotlin/org/rust/lang/core/resolve/ref/RsMethodCallReferenceImpl.kt [hunk 3]
```diff
@@ -31,7 +31,12 @@ class RsMethodCallReferenceImpl(
         val lookup = ImplLookup.relativeTo(element)
         val receiver = element.receiver.type
         return collectCompletionVariants {
-            processMethodCallExprResolveVariants(lookup, receiver, filterMethodCompletionVariants(it, lookup, receiver))
+            processMethodCallExprResolveVariants(lookup, receiver,
+                filterCompletionVariantsByVisibility(
+                    filterMethodCompletionVariants(it, lookup, receiver),
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/resolve/ref/RsMethodCallReferenceImpl.kt]: Maybe now we should rename `filterMethodCompletionVariants` to something like `filterMethodCompletionVariantsByTraitBounds` ?

**Comment by Undin** [comment]
Also, it would be great to have tests when we can't resolve some item or when we don't get some item to completion list because it's private

**Pull request merged**